private/bufpkg/bufimage/source_retention_options.go

```diff
@@ -0,0 +1,400 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufimage
+
+import (
+	"fmt"
+
+	imagev1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/image/v1"
+	"github.com/bufbuild/buf/private/pkg/protodescriptor"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// StripSourceRetentionOptionsFromFile returns a file descriptor proto that omits any
+// options that in file that are defined to be retained only in source. If file has no
+// such options, then it is returned as is. If it does have such options, a copy is
+// made; the given file will not be mutated.
+//
+// The given file must be either a *descriptorpb.FileDescriptorProto or a
+func StripSourceRetentionOptionsFromFile[F protodescriptor.FileDescriptor](
```

Comment:
I'm already finding usage for this to construct `CodeGeneratorRequests` outside of this repository (specifically for `protoplugin` testing) - what do you think about moving a `FileDescriptorProto`-specific version of this to say `protocompile` in some capacity? I'm not sure if `protocompile` is the right place (`protobuf-go` is really the right place), but just somewhere else that we can consume it.

Then, I would have a private function here that takes a `*imagev1.Image`, calls the public function for `FileDescriptorProto`, and then uses `fileDescriptorProtoToProtoImageFile` (which is in this package).

Reply:
Yeah, I think `protocompile` makes sense. Computing this stuff is the job of a compiler anyhow (though it's usually the code-gen part of the compiler, whereas protocompile only produces intermediate representations).

It probably makes the most sense inside the options sub-package there.

Reply:
https://github.com/bufbuild/protocompile/pull/250

Reply:
This function needs to be updated post-250, but I assume thats just coming up next

Reply:
Yeah, there now.